### PR TITLE
Add `BigInt64Array` and `BigUint64Array` tests

### PR DESCRIPTION
**Pull request by ExE-Boss**
**Spec:** <https://github.com/heycam/webidl/pull/936>

---

/cc @annevk @bathos

### CODE REVIEW

**Comment by annevk** [comment]
I think we should tests this in a couple other places as well, to ensure implementations end up supporting this correctly (or already support it correctly?):

* FileAPI/blob/Blob-constructor.any.js
* encoding/encodeInto.any.js
* fetch/api/response/response-clone.any.js
* html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/window-simple-success.https.html
* xhr/send-data-sharedarraybuffer.any.js

**Comment by annevk** [comment]
Thanks, I'm not a big fan of making support conditional. Is that needed for some implementations?

File: encoding/encodeInto.any.js [hunk 1]
```diff
@@ -127,11 +127,14 @@
  Uint16Array,
  Uint32Array,
  Uint8ClampedArray,
+ typeof BigInt64Array === "function" ? BigInt64Array : undefined,
+ typeof BigUint64Array === "function" ? BigUint64Array : undefined,
```
Comment (ExE-Boss) [inline encoding/encodeInto.any.js:131]: The `typeof` operator is needed to not cause a `ReferenceError` to be thrown while setting up the tests.

File: fetch/api/response/response-clone.any.js [hunk 2]
```diff
@@ -119,6 +119,8 @@ testReadableStreamClone(new Uint8Array(arrayBuffer), "Uint8Array");
 testReadableStreamClone(new Uint8ClampedArray(arrayBuffer), "Uint8ClampedArray");
 testReadableStreamClone(new Uint16Array(arrayBuffer, 2), "Uint16Array");
 testReadableStreamClone(new Uint32Array(arrayBuffer), "Uint32Array");
+testReadableStreamClone(typeof BigInt64Array === "function" ? new BigInt64Array(arrayBuffer) : undefined, "BigInt64Array");
+testReadableStreamClone(typeof BigUint64Array === "function" ? new BigUint64Array(arrayBuffer) : undefined, "BigUint64Array");
```
Comment (ExE-Boss) [inline fetch/api/response/response-clone.any.js:123]: ditto

File: encoding/encodeInto.any.js (hunk 1 above)
Comment (annevk) [reply encoding/encodeInto.any.js:131]: This is exposed in all browsers. Can you give some more detail as to where this goes wrong?

Comment (ExE-Boss) [reply encoding/encodeInto.any.js:131]: `BigInt64Array` and `BigUint64Array` are not supported in **Safari 14**, only in **Safari Technical Preview**.

**Comment by bathos** [comment]
I think the first place I encountered missing support was `crypto.getRandomValues`. However it’s unclear to me whether it should or shouldn’t work even after these changes now that I’m looking at the spec. Perhaps one of you will find this clearer than I do.

The argument type for `getRandomValues` is `ArrayBufferView`, but the [operation steps][1] begin with this step:

> 1. If array is not of an integer type (i.e., Int8Array, Uint8Array, Int16Array, Uint16Array, Int32Array, Uint32Array or Uint8ClampedArray), throw a TypeMismatchError and terminate the algorithm.

I haven’t found anything defining what it means for an ArrayBufferView to be “of an integer type.” Web IDL does define “integer type,” but TypedArrays are ES objects whose indexed members model ES Number or BigInt language values. They have associated constraining [“element types”][2] (e.g. Int8), but despite the general alignment (e.g. Int8 → Byte), these aren’t really Web IDL integer values, so it doesn’t seem like the Web IDL term would be applicable here?

Stepping back from muddling through a search for a formal answer about how to read this, though: does anybody know what the _intention_ might have been? At the time this was written, I’d guess they specifically wished to forbid Float32Array and Float64Array ... and ... maybe forgot DataView exists? I dunno.

If “of an integer type” really isn’t defined, I’d like to say big ones count (the most integer type of all, perhaps!) — but w/o knowing the rationale for the original constraint, I don’t feel confident making a case for that.

[1]: https://w3c.github.io/webcrypto/#Crypto-method-getRandomValues
[2]: https://tc39.es/ecma262/#sec-typedarray-objects

**Comment by annevk** [comment]
@bathos that specification seems woefully imprecise. I recommend filing an issue. It does seem like something that ought to work.

File: encoding/encodeInto.any.js (hunk 1 above)
Comment (annevk) [reply encoding/encodeInto.any.js:131]: Honestly I think that's fine for these tests. That's also what all the tooling uses.

Comment (ExE-Boss) [reply encoding/encodeInto.any.js:131]: That would lead to a **Harness status** of `FAIL`, and any tests defined after these lines to be `MISSING`.

Comment (annevk) [reply encoding/encodeInto.any.js:131]: But only in Safari 14, right? And all the bots use Safari TP? I feel like I'm missing something.

Comment (ExE-Boss) [reply encoding/encodeInto.any.js:131]: **Apple** likely needs to run these tests on **Safari 14** while it’s still a supported version.

Comment (annevk) [reply encoding/encodeInto.any.js:131]: I guess it's reasonable to work around that, but the amount of tests should not end up varying as it does now. In that case just quote them all and use `new globalThis[viewAsString](...)` or some such.

**Review by TimothyGu — APPROVED** [review]

**Comment by TimothyGu** [comment]
Hi @web-platform-tests/wpt-core-team. I'd like to merge this PR, which changes idlharness and thus potentially affects a lot of tests. However, the stability bots are unhappy about some pre-existing flakes:

- Firefox: `/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/blob-data.https.html`
- Chrome: `/webtransport/idlharness.any.html`

Is there a way to merge the PR anyway, despite the required statuses failing?

**Comment by foolip** [comment]
@TimothyGu yes if you've looked into the flakiness and it's pre-existing / unrelated then we can admin merge and bypass requires CI checks. I'll do just that.

**Pull request merged**